HerPublicWebsite.BusinessLogic/Services/CsvFileCreator/CsvFileCreator.cs

```diff
@@ -10,26 +10,149 @@ namespace HerPublicWebsite.BusinessLogic.Services.CsvFileCreator;
     {
-        var rows = referralRequests.Select(rr => new CsvRow(rr));
+        var rows = referralRequests.Select(rr => new CsvRowReferralRequest(rr));
+        return GenerateCsvMemoryStreamFromFileRows(rows);
+    }
 
-        var csvConfiguration = new CsvConfiguration(CultureInfo.InvariantCulture)
-        {
-            InjectionOptions = InjectionOptions.Strip
-        };
+    public MemoryStream CreateReferralRequestOverviewFileData(IEnumerable<ReferralRequest> referralRequests)
+    {
+        var rows = referralRequests.Select(rr => new CsvRowReferralCodes(rr));
+        return GenerateCsvMemoryStreamFromFileRows(rows);
+    }
 
-        using var writeableMemoryStream = new MemoryStream();
-        using var streamWriter = new StreamWriter(writeableMemoryStream, Encoding.UTF8);
-        using var csvWriter = new CsvWriter(streamWriter, csvConfiguration);
-        {
-            csvWriter.WriteRecords(rows);
-            csvWriter.Flush();
-            return new MemoryStream(writeableMemoryStream.GetBuffer(), 0, (int)writeableMemoryStream.Length, false);
+    public MemoryStream CreateReferralRequestFollowUpFileData(IEnumerable<ReferralRequest> referralRequests)
+    {
+        var rows = referralRequests
+            .GroupBy(rr => rr.CustodianCode)
+            .GroupBy(groupingByLa =>
+                LocalAuthorityData.LocalAuthorityDetailsByCustodianCode[groupingByLa.Key].Consortium)
+            .SelectMany(groupingByConsortium =>
+                {
+                    var consortiumReferrals = groupingByConsortium.SelectMany(g => g).ToList();
+                    var consortiumStatistics = new ConsortiumStatistics(consortiumReferrals);
+                    return groupingByConsortium.Select(groupingByLa => new CsvRowLaDownloadInformation(groupingByLa, consortiumStatistics));
+                }
+            );
+        return GenerateCsvMemoryStreamFromFileRows(rows);
+    }
+
+    private class CsvRowReferralCodes
+    {
+        [Index(0)]
+        [Name("Consortium")]
+        public string Consortium { get; set; }
+        
+        [Index(1)]
+        [Name("Local Authority")]
+        public string LocalAuthority { get; set; }
+        
+        [Index(2)]
+        [Name("Referral Code")]
+        public string ReferralCode { get; set; }
+        public CsvRowReferralCodes(ReferralRequest request){
+            Consortium =  LocalAuthorityData.LocalAuthorityDetailsByCustodianCode[request.CustodianCode].Consortium;
+            LocalAuthority =  LocalAuthorityData.LocalAuthorityDetailsByCustodianCode[request.CustodianCode].Name;
+            ReferralCode = request.ReferralCode;
+        }
+    }
+    
+    private class ConsortiumStatistics
+    {
+        public bool AllConsortiumReferralsDownloaded { get; set; }
+        public int NumberUndownloadedConsortiumReferrals { get; set; }
+        public float PercentageUndownloadedConsortiumReferrals { get; set; }
```

Comment:
I'd usually stick with `double` unless the lower precision is particularly important.

I'm also wondering if they'll want the percentages to be rounded (i.e. so it doesn't display a massive decimal), but that's probably not MVP. Keeping it as an int would be a simple way of doing so.

Reply:
Good point, I'll swap it to double.

I'm not convinced there's value to rounding it. I assume if anyone is reading the data by eye, they'll use something like Excel and so can format the data there. If entering the data somewhere it's probably best to have it as accurate as possible.